mm2src/mm2_main/src/lp_ordermatch.rs

```diff
@@ -2049,31 +2075,66 @@ impl MakerOrder {
             return OrderMatchResult::NotMatched;
         }
 
+        let is_legacy = self.swap_version.is_legacy() || taker.swap_version.is_legacy();
+
         match taker.action {
             TakerAction::Buy => {
                 let ticker_match = (self.base == taker.base
                     || self.base_orderbook_ticker.as_ref() == Some(&taker.base))
                     && (self.rel == taker.rel || self.rel_orderbook_ticker.as_ref() == Some(&taker.rel));
+                // taker_base_amount: the amount taker desires to buy (input.volume from SellBuyRequest)
+                // taker_rel_amount: the amount  taker is willing to pay (input.volume * input.price, where input is SellBuyRequest)
+                // taker_price: the effective price offered by the taker
                 let taker_price = taker_rel_amount / taker_base_amount;
-                if ticker_match
-                    && taker_base_amount <= &self.available_amount()
-                    && taker_base_amount >= &self.min_base_vol
-                    && taker_price >= self.price
-                {
+
+                // Reject if any basic conditions are not satisfied
+                let base_amount_exceeds = taker_base_amount > &self.available_amount();
+                let below_min_volume = taker_base_amount < &self.min_base_vol;
+                let price_too_low = taker_price < self.price;
+
+                if !ticker_match || base_amount_exceeds || below_min_volume || price_too_low {
+                    return OrderMatchResult::NotMatched;
+                }
+
+                if is_legacy {
+                    // Legacy mode: use maker's price to calculate rel amount
                     OrderMatchResult::Matched((taker_base_amount.clone(), taker_base_amount * &self.price))
                 } else {
-                    OrderMatchResult::NotMatched
+                    // taker_rel_amount must cover the premium requested by maker
+                    let required_rel_amount =
+                        taker_base_amount * &self.price + self.premium.clone().unwrap_or_default();
```

Comment:
Here we require Taker to pay Maker order's premium.
As I can see, premium is a fixed amount in rel coin units, assigned when the order is created.
What if the order is filled only partially, will Taker always pay the same premium amount (that is, shouldn't premium be proportional to the order amount being filled)?

Reply:
> What if the order is filled only partially, will Taker always pay the same premium amount 

yes, premium is a fixed rel amount maker requires for a swap. They ask it for the action (swap) not for a trading volume.

Reply:
I think this is not fair:
If an order was filled 3 times partially the maker would receive 3 premiums.
If same order is filled only one time for the full amount the maker receives only 1 premium.
This seems unfair. I think the maker should receive same premium no matter how many times the order was filled.
For this the premium should be calculated proportionally to the amount filled.

(Not sure though, maybe this was discussed already and was decided to pay premium like it is implemented now)

Reply:
During swap maker pays fees for transactions, which doesn't depend from tx value, its better to treat premium as guaranteed fixed amount which could cover maker expenses on swap